sycl/doc/SYCLGraphUsageGuide.md

```diff
@@ -0,0 +1,493 @@
+accessing the same memory resources. However, it is quite common to wish to
+execute the same graph on different sets of data concurrently. This can be
+accomplished by using graph update functionality, but this must also be
+serialized with a host synchronization. Consider this example (it uses whole
+graph update for simplicity):
+
+```c++
+sycl::command_graph<graph_state::executable> ExecutableGraph
+   = ModifiableGraph.finalize();
+
+Queue.ext_oneapi_graph(ExecutableGraph);
+// Updating the graph here to use new memory, this forces a host synchronization
+ExecutableGraph.update(ModifiableGraphWithNewParams);
+
+// Re-execute the update graph
+Queue.ext_oneapi_graph(ExecutableGraph);
+
+```
+
+If your sets of inputs are known (such as a double-buffering type scenario), you
+can create multiple executable graphs up front, avoiding the need for expensive
+host sync between executions (assuming there are no implicit dependencies from
+accessing the same `sycl::buffer`) for updating, and potentially increasing
+device occupancy (if the device supports it). Modifying the example from above:
+
+```c++
+sycl::command_graph<graph_state::executable> ExecutableGraph
+   = ModifiableGraph.finalize();
+sycl::command_graph<graph_state::executable> ExecutableGraphOtherParams
+   = ModifiableGraphWithNewParams.finalize();
+
+Queue.ext_oneapi_graph(ExecutableGraph);
+// We can now execute the second graph with no implicit dependency
+Queue.ext_oneapi_graph(ExecutableGraphOtherParams);
+
+// Repeatedly executing the graphs will create dependencies on their individual
+// preceeding executions, but not on each other.
+Queue.ext_oneapi_graph(ExecutableGraph);
+Queue.ext_oneapi_graph(ExecutableGraphOtherParams);
+
+```
+
+If the amount of different inputs is not always known up front a similar
+strategy could be employed but instead using a pool of graphs to hide the
+potential host-synchronization caused when updating and increase device
+occupancy.
+
+
+### Recording Library Calls
+
+#### Warmups
+
+Some libraries may perform one-time warmups on first execution of some
+operations (for example to trigger kernel creation) that are not desirable to
+capture in a graph. Consider the following simplified example:
+
+```c++
+// Hypothetical library function
+void SomeLibrary::Operation(sycl::Queue Queue){
+static bool IsFirstTime = true;
+
+if(IsFirstTime){
+  // Warmup by launching kernel once
+  do_warmup(Queue);
+  // Execute the actual operation
+  execute_operation(Queue);
+  IsFirstTime = false;
+} else{
+  execute_operation(Queue);
+}
+}
+
+// SYCL Application Code
+
+Graph.begin_recording(Queue);
+
+// do_warmup() will be captured here and executed every time the graph is
+// executed in future which is undesirable.
+SomeLibrary::Operation(Queue);
+
+Graph.end_recording(Queue);
+
+```
+
+In this case it may be necessary to first manually trigger the warmup by calling
+`SomeLibrary::Operation()` to prevent the warmup from being captured in a graph
```

Comment:
```suggestion
`SomeLibrary::Operation()` before starting to record the queue with `Graph.begin_recording(Queue)` to prevent the warmup from being captured in a graph
```